lib/tree.rb

```diff
@@ -16,40 +16,101 @@ def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(log n)
+  # Space Complexity: O(1)
+  def add_recursive(node, key, value)
+    return TreeNode.new(key, value) if node.nil?
+    if key < node.key
+      node.left = add_recursive(node.left, key, value)
+    else
+      node.right = add_recursive(node.right, key, value)
+    end
+    return node
+  end
+
   def add(key, value)
-    raise NotImplementedError
+    @root = add_recursive(@root, key, value)
+  end
+
+  # Time Complexity: O(log n)
+  # Space Complexity: O(1)
+  def find_recursive(node, key)
```

Comment:
👍 , However the space complexity is O(log n) if the tree is balanced because of the call stack.